src/react/utils/useHydrateAtoms.ts

```diff
@@ -4,11 +4,20 @@ import type { WritableAtom } from '../../vanilla.ts'
 type Store = ReturnType<typeof useStore>
 type Options = Parameters<typeof useStore>[0]
 type AnyWritableAtom = WritableAtom<unknown, any[], any>
+type AtomTuple<A = AnyWritableAtom, V = unknown> = readonly [A, V]
+
+type InferAtoms<T extends Iterable<AtomTuple>> = {
+  [K in keyof T]: T[K] extends AtomTuple<infer A>
+    ? A extends AnyWritableAtom
+      ? AtomTuple<A, ReturnType<A['read']>>
+      : T[K]
+    : any
```

Comment:
If this library drops support for `Map` this could be `never`. Feels less safe to allow `any`, not sure what else to do for continued support for passing `Map`.

Reply:
For now, I'd like to keep supporting `Map`s.